✨[#1] 경매 상품 등록

puclpu opened the pull request:
## #️⃣연관된 이슈
- 이슈 번호: #1 

## 작업 내용
- 버킷 연결
    - dependency 추가
    - application.yaml에 설정 추가
- 상품 등록 api 구현
    - 경매 상품 저장
    - 상품 이미지 업로드 및 저장
    - 등록된 경매 상품 정보 반환
    - 이미지 업로드 과정에서 발생하는 예외 처리

## 체크리스트
- [x] 코드가 제대로 동작하는지 확인했습니다.
- [x] 문서(코드, 주석, README 등)를 업데이트했습니다.

puclpu committed:
➕ [#1] S3 버킷 연결

- spring cloud aws starter 의존성 추가
- application.yaml 에 s3 설정 추가
- S3Config 설정 파일 추가

puclpu committed:
✨ [#1] 경매 상품 등록 기능

- 경매 상품 저장
- 경매 상품 이미지 업로드 및 저장
- 등록된 경매 상품 정보 반환
- 이미지 업로드 과정 시 발생하는 예외 처리

je-pa commented on goodsending/src/main/java/com/goodsending/product/dto/request/ProductCreateRequestDto.java:
```diff
@@ -0,0 +1,24 @@
+@Getter
+public class ProductCreateRequestDto {
+
+  @Size(min = 1, max = 20, message = "상품명은 1글자 이상 20자 이하로 입력할 수 있습니다.")
+  private String name;
+
+  @Min(value = 0, message = "판매가를 0원 이상 입력해야합니다.")
+  private int price;
+
+  @Size(min = 1, max = 255, message = "상품 소개는 1글자 이상 255자 이하로 입력할 수 있습니다.")
+  private String introduction;
+
+  @Min(value = 1, message = "마감 기한은 1일 이상이어야 합니다.")
+  @Max(value = 3, message = "마감 기한은 3일 이하여야 합니다.")
+  private int auctionPeriodDays;
```
null값은 못 잡아줄 것 같아요!

puclpu replied:
혹시 int 부분을 말씀하시는걸까요?

je-pa replied:
int는 괜찮을 것 같은데 Size도 null 체크 안해주지 않나요?

puclpu replied:
NotNull Bean Validation 추가해주었습니다! 감사합니다

je-pa commented on goodsending/src/main/java/com/goodsending/product/entity/Product.java:
```diff
@@ -0,0 +1,70 @@
+  public static Product of(ProductCreateRequestDto requestDto, Long memberId) {
+
+    int auctionPeriodDays = requestDto.getAuctionPeriodDays();
+    LocalDateTime now = LocalDateTime.now();
```
now 같은 직접 제어할 수 없는 값은 상위 계층으로 분리하는 것이 좋다고 하더라고요!
controller부터 생성해서 파라미터로 전달해주는게 단위 테스트할 때도 편하다고 합니다.
또한 now가 여러군데 사용된다면 하나의 now로 통일해줄 수 도 있다고 합니다!

puclpu replied:
global/service 패키지에 TimeService 클래스를 생성하여 현재 시각을 가져오는 로직을 옮겨주었습니다. 감사합니다.

puclpu replied:
Controller에서 현재 시각 객체를 생성하여 Service에 파라미터로 전달하는 방식으로 수정하였습니다.

je-pa commented on goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+  private final S3Uploader s3Uploader;
+
+  @Override
+  public ProductCreateResponseDto createProduct(ProductCreateRequestDto requestDto,
```
상품과 상품이미지를 저장하는데 둘중 하나만 저장되면 안되니까 Transactional 이 필요하지 않을까요?

puclpu replied:
수정했습니다!

je-pa commented on goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+    // 상품 이미지 업로드
+    List<ProductImageInfoDto> savedProductImages = new ArrayList<>();
+    for (MultipartFile productImageFile : productImages) {
```
요부분은 동기적으로 처리가 될까요?

je-pa replied:
List<MultipartFile> productImages 자체를 파라미터로 받는 메서드를 s3업로더 클래스에 만드시는 것도 좋을 것 같아요

puclpu replied:
productImages를 파라미터로 받는 메서드 uploadProductImageFileList를 생성하여 업로드 된 파일의 url 목록을 받아오는 방식으로 변경하였습니다. 감사합니다.

je-pa commented on goodsending/src/main/java/com/goodsending/product/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    // 상품 이미지 업로드
+    List<ProductImageInfoDto> savedProductImages = new ArrayList<>();
+    for (MultipartFile productImageFile : productImages) {
+      try {
+        // S3 업로드
+        String storedFileName = s3Uploader.upload(productImageFile, "images/products");
+
+        // 업로드 된 url 저장
+        ProductImage productImage = ProductImage.of(product, storedFileName);
+        productImageRepository.save(productImage);
+
+        ProductImageInfoDto productImageInfoDto = ProductImageInfoDto.from(productImage);
+        savedProductImages.add(productImageInfoDto);
+      } catch (MaxUploadSizeExceededException e) { // 파일 용량 초과
+        throw CustomException.of(ExceptionCode.FILE_SIZE_EXCEEDED);
+      } catch (IOException e) {
+        if (e.getMessage().contains("No space left on device")) {
+          throw CustomException.of(ExceptionCode.LOW_DISK_SPACE); // 디스크 공간 부족
+        } else {
+          throw CustomException.of(ExceptionCode.FILE_UPLOAD_FAILED); // 파일 업로드 실패
+        }
+      }
```
try문안에는 예외가 일어날수 있는 부분의 최소한의 코드를 작성해주시는게 좋을 것 같아요!

try catch문 service에서 말고 uploader 에서 처리해주면 나중에 해당 메서드를 사용할때 중복적으로 예외처리를 안해주어도 될 것 같아요.

puclpu replied:
S3Uploader의 convert 메서드에서 예외처리하는 방식으로 변경했습니다. 감사합니다.

je-pa reviewed (commented):
고생하셨습니다:)

puclpu committed:
🐛 [#1] 경매 상품 등록 기능 Transactional 추가

- ProductServiceImpl 의 경매 상품 등록 기능 Transactional 추가

puclpu committed:
♻️ [#1] 경매 상품 이미지 업로드 로직 변경

- 업로드 된 파일의 url 목록을 받아 상품 이미지 객체를 생성하여 List에 담는 방식으로 변경
- 예외 처리를 IOException이 발생하는 convert 메서드에서 처리

puclpu committed:
♻️ [#1] 현재 시작 생성 로직 변경

- now 메서드를 통해 현재 시각을 가져오는 로직을 TimeService 를 생성하여 구현

puclpu committed:
🐛 [#1] 상품 등록 요청 Dto Null 체크

- ProductCreateRequestDto 의 상품명, 상품 소개 필드에 null 체크 Bean Validation 추가

puclpu committed:
🔥️ [#1] TimeService 삭제

puclpu committed:
♻️ [#1] 등록 시각 생성 로직 변경

- ProductController 에서 currentTime 생성

puclpu committed:
🐛 [#1] 상품 등록 반환 dto 파라미터 수정

- product에서 savedProduct로 수정
